internal/gcs/container.go

```diff
@@ -0,0 +1,171 @@
+package gcs
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/Microsoft/hcsshim/internal/cow"
+	"github.com/Microsoft/hcsshim/internal/logfields"
+	"github.com/Microsoft/hcsshim/internal/schema1"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	hrComputeSystemDoesNotExist = 0xc037010e
+)
+
+func (gc *GuestConnection) requestNotify(cid string, ch chan struct{}) error {
+	gc.mu.Lock()
+	defer gc.mu.Unlock()
+	if gc.notifyChs == nil {
+		return errors.New("guest connection closed")
+	}
+	if _, ok := gc.notifyChs[cid]; ok {
+		return fmt.Errorf("container %s already exists", cid)
+	}
+	gc.notifyChs[cid] = ch
```

Comment:
I assume 1 notify per `cid` is ok?

Reply:
I believe that's all that's defined so far.

Reply:
Yea as long as you don't use any "open" workflows. But I think that's a safe assumption at this point.

Reply:
I don't think there are any open workflows at the bridge protocol level; I think HCS is multiplexing the notification to support open.